`terraform apply` crashes if a variable's default is set to a list

If you set a variable’s default to a list, then `terraform apply` exits with an ugly crash.

Here’s my test case:

1. Make a new directory and cd into it:

    ```shell
    $ cd $(mktemp -d -t terraformtest)
    ```
2. Make a testcase terraform config:
    ```shell

$ cat > test.tf  << EOF
variable "testcase" {
  default {
    l = [1, 2, 3]
  }
}
EOF
    ```
3. Try running terraform apply:
    ```shell
$ terraform apply
panic: Unknown variable type: 0x0

goroutine 1 [running]:
github.com/hashicorp/terraform/terraform.(*Context).Input(0xc208094b00, 0xc208094b07, 0x0, 0x0)
 /Users/spencer/go/src/github.com/hashicorp/terraform/terraform/context.go:175 +0xc61
github.com/hashicorp/terraform/command.(*ApplyCommand).Run(0xc2080641c0, 0xc20800a010, 0x0, 0x0, 0xc2080368a8)
 /Users/spencer/go/src/github.com/hashicorp/terraform/command/apply.go:125 +0x11a6
github.com/mitchellh/cli.(*CLI).Run(0xc20806a500, 0xc2080d8ae0, 0x0, 0x0)
 /Users/spencer/go/src/github.com/mitchellh/cli/cli.go:100 +0x3ce
main.wrappedMain(0x0)
 /Users/spencer/go/src/github.com/hashicorp/terraform/main.go:150 +0xb76
main.realMain(0x0)
 /Users/spencer/go/src/github.com/hashicorp/terraform/main.go:82 +0x923
main.main()
 /Users/spencer/go/src/github.com/hashicorp/terraform/main.go:18 +0x1f

goroutine 5 [syscall]:
os/signal.loop()
 /usr/local/Cellar/go/1.4.2/libexec/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
 /usr/local/Cellar/go/1.4.2/libexec/src/os/signal/signal_unix.go:27 +0x35

goroutine 8 [chan receive]:
main.func·014()
 /Users/spencer/go/src/github.com/hashicorp/terraform/commands.go:134 +0x52
created by main.makeShutdownCh
 /Users/spencer/go/src/github.com/hashicorp/terraform/commands.go:137 +0x1b4

goroutine 9 [semacquire]:
sync.(*Cond).Wait(0xc2080d4330)
 /usr/local/Cellar/go/1.4.2/libexec/src/sync/cond.go:62 +0x9e
io.(*pipe).read(0xc2080d4300, 0xc20809f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
 /usr/local/Cellar/go/1.4.2/libexec/src/io/pipe.go:52 +0x303
io.(*PipeReader).Read(0xc2080441b0, 0xc20809f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
 /usr/local/Cellar/go/1.4.2/libexec/src/io/pipe.go:134 +0x5b
bufio.(*Scanner).Scan(0xc208066300, 0x0)
 /usr/local/Cellar/go/1.4.2/libexec/src/bufio/scan.go:180 +0x688
github.com/hashicorp/terraform/command.func·006()
 /Users/spencer/go/src/github.com/hashicorp/terraform/command/meta.go:293 +0x3d
created by github.com/hashicorp/terraform/command.(*Meta).flagSet
 /Users/spencer/go/src/github.com/hashicorp/terraform/command/meta.go:296 +0x51f



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at “crash.log” relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
    ```

I stumbled on this while crossing my fingers that I could set a default to a list of CIDR blocks to whitelist in an AWS security group. I’m fine with lists being verboten, but the error should be clearer.

--- a/terraform/context.go
+++ b/terraform/context.go
@@ -175,6 +175,8 @@ func (c *Context) Input(mode InputMode) error {
 
 			v := m[n]
 			switch v.Type() {
+			case config.VariableTypeUnknown:
+				continue
 			case config.VariableTypeMap:
 				continue
 			case config.VariableTypeString:

--- a/terraform/context_test.go
+++ b/terraform/context_test.go
@@ -2883,6 +2883,28 @@ func TestContext2Input(t *testing.T) {
 	}
 }
 
+func TestContext2Input_badVarDefault(t *testing.T) {
+	m := testModule(t, "input-bad-var-default")
+	p := testProvider("aws")
+	p.ApplyFn = testApplyFn
+	p.DiffFn = testDiffFn
+	ctx := testContext2(t, &ContextOpts{
+		Module: m,
+		Providers: map[string]ResourceProviderFactory{
+			"aws": testProviderFuncFixed(p),
+		},
+	})
+
+	p.InputFn = func(i UIInput, c *ResourceConfig) (*ResourceConfig, error) {
+		c.Config["foo"] = "bar"
+		return c, nil
+	}
+
+	if err := ctx.Input(InputModeStd); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+}
+
 func TestContext2Input_provider(t *testing.T) {
 	m := testModule(t, "input-provider")
 	p := testProvider("aws")

--- a/terraform/test-fixtures/input-bad-var-default/main.tf
+++ b/terraform/test-fixtures/input-bad-var-default/main.tf
@@ -1,0 +1,5 @@
+variable "test" {
+    default {
+        l = [1, 2, 3]
+    }
+}

